### ISSUE TITLE
Django's use of asgiref/sync.py

### ISSUE THREAD

**Issue by a-feld** [django/asgiref#223]
Hi @andrewgodwin , thanks for all your help on this library and Django! At @newrelic we've been working on adding automated instrumentation support for Django's ASGI mode.

Django internally uses asgiref's `SyncToAsync` and `AsyncToSync` implementations. We've been struggling to figure out what the intended behavior is for Django's use case with `thread_sensitive`.

According to the [docs](https://docs.djangoproject.com/en/3.1/topics/async/) the intent is to have it per-request:
>  ...then Django must use a thread per request to safely emulate a synchronous environment for it.

However, the way the asgiref library is used in django results in the `single_thread_executor` being scoped to the process, meaning that it's 1 thread _total_.

It looks to me like asgiref is intentional in having only 1 thread pool **per process** based on the tests, but I'm thinking Django might be resilient to multi-threading as long as all sync work is scheduled on the same thread (1 thread **per request**).

I have a patch to change over to a request-specific single thread pool that I'm happy to contribute, but I don't want to ship it up without first understanding if this is the intent in both asgiref & Django.

### TL/DR :)
In short:
* Is the intent of Django to run 1 sync thread **per request** or **per process**?
* Is the intent of the `thread_sensitive` parameter to be 1 thread **per process** or **per asyncio root task**?

#### urls.py
```python
import time

from django.http import HttpResponse
from django.urls import path

def view(request):
    print("STARTING")
    time.sleep(10)
    return HttpResponse("*")

urlpatterns = [
    path("", view),
]
```

#### settings.py
```python
ROOT_URLCONF = "urls"
SECRET_KEY = "shhh"
DEBUG = True
```

```bash
daphne asgi:application &
curl localhost:8000 &
curl localhost:8000
```

I'm wondering if with this example app, we expect it to run for 10 seconds or 20 seconds?

**Comment by andrewgodwin** [comment]
The intention of `thread_sensitive` is to run one thread per request - however, this is implemented as one thread per process currently, as that was the safest and most compatible way of achieving this. Obviously, this will be relatively OK with multi-process worker scenarios and not great with multi-threaded ones.

If you have a patch for making it per-request I'm more than happy to look at it, but I want to make sure we do it in a safe way (that is, asgiref is in one-thread-per-process mode unless something comes along and gives it signalling that allows one-thread-per-request). You might also need to patch the use of `sync_to_async` in the AsgiHandler to get things working fully, but we can separate that out from just getting this working in the general multi-thread case.

**Comment by a-feld** [comment]
Thanks for the reply! I agree with the safety-first approach, this stuff is very tricky! I've already spotted a couple of bugs in my original implementation. I'll have to spend some more time working through this. Thanks again!
